File: arrays_and_strings/combine.cpp
#include <gtest/gtest.h>

#include <algorithm>
#include <vector>

static std::vector<int> combine(std::vector<int> arr1, std::vector<int> arr2)
{
    //! @details Time and space complexity O(arr1.size() + arr2.size())
    //!          Space complexity constant if don't count output as extra space
    
    std::vector<int> ans {};

    int i = 0, j = 0;

    while (i < arr1.size() && j < arr2.size())
    {
        if (arr1[i] < arr2[j])
        {
            ans.push_back(arr1[i]);
            ++i;
        }
        else
        {
            ans.push_back(arr2[j]);
            ++j;
        }
    }
    
    while (i < arr1.size())
    {
        ans.push_back(arr1[i]);
        ++i;
    }
    
    while (j < arr2.size())
    {
        ans.push_back(arr2[j]);
        ++j;
    }

    return ans;

} // static std::vector<int> combine( ...

TEST(CombineTest, SampleTest)
{
    const std::vector<int> expected_output {1, 3, 4, 5, 6, 7, 20};

    const auto result = combine({1, 4, 7, 20}, {3, 5, 6});

    EXPECT_TRUE(std::equal(expected_output.cbegin(),
                           expected_output.cend(),
                           result.cbegin()));
}

File: arrays_and_strings/combine.java
public List<Integer> combine(int[] arr1, int[] arr2)
{
    List<Integer> ans = new ArrayList<>();
    int i = 0;
    int j = 0;

    while (i < arr1.length && j < arr2.length)
    {
        if (arr1[i] < arr2[j])
        {
            ans.add(arr1[i]);
            i++;
        }
        else
        {
            ans.add(arr2[j]);
            j++;
        }
    }

    while (i < arr1.length)
    {
        ans.add(arr1[i]);
        i++;
    }

    while (j < arr2.length)
    {
        ans.add(arr2[j]);
        j++;
    }

    return ans;
}